UNION query fails for UnknownType

With configuration

    optimizer.optimize-hash-generation=true

This query fails

    (VALUES NULL) UNION (VALUES NULL)

Stack trace

    com.facebook.presto.spi.PrestoException: Compiler failed and interpreter is disabled
    	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitScanFilterAndProject(LocalExecutionPlanner.java:909)
    	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitProject(LocalExecutionPlanner.java:808)
    	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitProject(LocalExecutionPlanner.java:410)
    	at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:81)
    	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitAggregation(LocalExecutionPlanner.java:705)
    	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitAggregation(LocalExecutionPlanner.java:410)
    	at com.facebook.presto.sql.planner.plan.AggregationNode.accept(AggregationNode.java:160)
    	at com.facebook.presto.sql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:242)
    	at com.facebook.presto.execution.SqlTaskExecution.<init>(SqlTaskExecution.java:164)
    	at com.facebook.presto.execution.SqlTaskExecution.createSqlTaskExecution(SqlTaskExecution.java:115)
    	at com.facebook.presto.execution.SqlTaskExecutionFactory.create(SqlTaskExecutionFactory.java:80)
    	at com.facebook.presto.execution.SqlTask.updateTask(SqlTask.java:229)
    	at com.facebook.presto.execution.SqlTaskManager.updateTask(SqlTaskManager.java:265)
    	at com.facebook.presto.server.TaskResource.createOrUpdateTask(TaskResource.java:103)
    	at sun.reflect.GeneratedMethodAccessor537.invoke(Unknown Source)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.lang.reflect.Method.invoke(Method.java:483)
    	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
    	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:151)
    	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:171)
    	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:152)
    	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:104)
    	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:387)
    	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:331)
    	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:103)
    	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:271)
    	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
    	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
    	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
    	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
    	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
    	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:297)
    	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:254)
    	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1030)
    	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:373)
    	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:381)
    	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:344)
    	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:221)
    	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:808)
    	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669)
    	at org.eclipse.jetty.servlets.UserAgentFilter.doFilter(UserAgentFilter.java:83)
    	at org.eclipse.jetty.servlets.GzipFilter.doFilter(GzipFilter.java:300)
    	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
    	at io.airlift.http.server.TraceTokenFilter.doFilter(TraceTokenFilter.java:62)
    	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
    	at io.airlift.http.server.TimingFilter.doFilter(TimingFilter.java:51)
    	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
    	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
    	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
    	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
    	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
    	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
    	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:110)
    	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
    	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:159)
    	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52)
    	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
    	at org.eclipse.jetty.server.Server.handle(Server.java:499)
    	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)
    	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
    	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
    	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
    	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
    	at java.lang.Thread.run(Thread.java:745)
    Caused by: com.google.common.util.concurrent.UncheckedExecutionException: java.lang.ArrayIndexOutOfBoundsException: -1
    	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2203)
    	at com.google.common.cache.LocalCache.get(LocalCache.java:3937)
    	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3941)
    	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4824)
    	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4830)
    	at com.facebook.presto.sql.gen.ExpressionCompiler.compilePageProcessor(ExpressionCompiler.java:87)
    	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitScanFilterAndProject(LocalExecutionPlanner.java:897)
    	... 63 more
    Caused by: java.lang.ArrayIndexOutOfBoundsException: -1
    	at org.objectweb.asm.Frame.a(Unknown Source)
    	at org.objectweb.asm.MethodWriter.visitMaxs(Unknown Source)
    	at com.facebook.presto.byteCode.MethodDefinition.visit(MethodDefinition.java:253)
    	at com.facebook.presto.byteCode.MethodDefinition.visit(MethodDefinition.java:207)
    	at com.facebook.presto.byteCode.ClassDefinition.visit(ClassDefinition.java:158)
    	at com.facebook.presto.sql.gen.CompilerUtils.defineClasses(CompilerUtils.java:100)
    	at com.facebook.presto.sql.gen.CompilerUtils.defineClass(CompilerUtils.java:74)
    	at com.facebook.presto.sql.gen.CompilerUtils.defineClass(CompilerUtils.java:80)
    	at com.facebook.presto.sql.gen.ExpressionCompiler.compileProcessor(ExpressionCompiler.java:130)
    	at com.facebook.presto.sql.gen.ExpressionCompiler.compileAndInstantiate(ExpressionCompiler.java:93)
    	at com.facebook.presto.sql.gen.ExpressionCompiler.access$300(ExpressionCompiler.java:42)
    	at com.facebook.presto.sql.gen.ExpressionCompiler$1.load(ExpressionCompiler.java:53)
    	at com.facebook.presto.sql.gen.ExpressionCompiler$1.load(ExpressionCompiler.java:48)
    	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3527)
    	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2319)
    	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2282)
    	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2197)
    	... 69 more

--- a/presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java
+++ b/presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java
@@ -102,6 +102,7 @@ import com.facebook.presto.type.TimeOperators;
 import com.facebook.presto.type.TimeWithTimeZoneOperators;
 import com.facebook.presto.type.TimestampOperators;
 import com.facebook.presto.type.TimestampWithTimeZoneOperators;
+import com.facebook.presto.type.UnknownOperators;
 import com.facebook.presto.type.VarbinaryOperators;
 import com.facebook.presto.type.VarcharOperators;
 import com.google.common.annotations.VisibleForTesting;
@@ -321,6 +322,7 @@ public class FunctionRegistry
                 .scalar(JsonFunctions.class)
                 .scalar(ColorFunctions.class)
                 .scalar(HyperLogLogFunctions.class)
+                .scalar(UnknownOperators.class)
                 .scalar(BooleanOperators.class)
                 .scalar(BigintOperators.class)
                 .scalar(DoubleOperators.class)

--- a/presto-main/src/main/java/com/facebook/presto/sql/gen/ByteCodeUtils.java
+++ b/presto-main/src/main/java/com/facebook/presto/sql/gen/ByteCodeUtils.java
@@ -247,7 +247,7 @@ public final class ByteCodeUtils
         else if (type == Void.class) {
             notNull.pushNull()
                     .checkCast(Void.class);
-            expectedCurrentStackType = void.class;
+            return notNull;
         }
         else {
             throw new UnsupportedOperationException("not yet implemented: " + type);

--- a/presto-main/src/main/java/com/facebook/presto/sql/gen/IsDistinctFromCodeGenerator.java
+++ b/presto-main/src/main/java/com/facebook/presto/sql/gen/IsDistinctFromCodeGenerator.java
@@ -22,6 +22,7 @@ import com.facebook.presto.metadata.OperatorType;
 import com.facebook.presto.metadata.Signature;
 import com.facebook.presto.spi.type.Type;
 import com.facebook.presto.sql.relational.RowExpression;
+import com.facebook.presto.type.UnknownType;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
 
@@ -58,6 +59,22 @@ public class IsDistinctFromCodeGenerator
                 .comment("equals(%s, %s)", leftType, rightType)
                 .append(invoke(binding, operator.getSignature()));
 
+        ByteCodeNode neitherSideIsNull;
+        if (leftType instanceof UnknownType || rightType instanceof UnknownType) {
+            // the generated block should be unreachable. However, a boolean need to be pushed to balance the stack
+            neitherSideIsNull = new Block()
+                    .comment("unreachable code")
+                    .push(false);
+        }
+        else {
+            // This code assumes that argument and return type are not @Nullable.
+            // It is not the case for UnknownType, and introduces Verification Error.
+            // And it is hard to imagine that making it work with @Nullable will be useful in any other cases;
+            neitherSideIsNull = new Block()
+                    .append(equalsCall)
+                    .invokeStatic(CompilerOperations.class, "not", boolean.class, boolean.class);
+        }
+
         Block block = new Block()
                 .comment("IS DISTINCT FROM")
                 .comment("left")
@@ -81,9 +98,7 @@ public class IsDistinctFromCodeGenerator
                                                 .pop(leftType.getJavaType())
                                                 .pop(rightType.getJavaType())
                                                 .push(true))
-                                        .ifFalse(new Block()
-                                                .append(equalsCall)
-                                                .invokeStatic(CompilerOperations.class, "not", boolean.class, boolean.class)))))
+                                        .ifFalse(neitherSideIsNull))))
                 .append(wasNull.set(constantFalse()));
 
         return block;

--- a/presto-main/src/main/java/com/facebook/presto/type/UnknownOperators.java
+++ b/presto-main/src/main/java/com/facebook/presto/type/UnknownOperators.java
@@ -1,0 +1,99 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.type;
+
+import com.facebook.presto.operator.scalar.ScalarOperator;
+import com.facebook.presto.spi.type.StandardTypes;
+
+import javax.annotation.Nullable;
+
+import static com.facebook.presto.metadata.OperatorType.BETWEEN;
+import static com.facebook.presto.metadata.OperatorType.EQUAL;
+import static com.facebook.presto.metadata.OperatorType.GREATER_THAN;
+import static com.facebook.presto.metadata.OperatorType.GREATER_THAN_OR_EQUAL;
+import static com.facebook.presto.metadata.OperatorType.HASH_CODE;
+import static com.facebook.presto.metadata.OperatorType.LESS_THAN;
+import static com.facebook.presto.metadata.OperatorType.LESS_THAN_OR_EQUAL;
+import static com.facebook.presto.metadata.OperatorType.NOT_EQUAL;
+
+public final class UnknownOperators
+{
+    private UnknownOperators()
+    {
+    }
+
+    @ScalarOperator(EQUAL)
+    @Nullable
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean equal(@SqlType("unknown") @Nullable Void left, @SqlType("unknown") @Nullable Void right)
+    {
+        return null;
+    }
+
+    @ScalarOperator(NOT_EQUAL)
+    @Nullable
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean notEqual(@SqlType("unknown") @Nullable Void left, @SqlType("unknown") @Nullable Void right)
+    {
+        return null;
+    }
+
+    @ScalarOperator(LESS_THAN)
+    @Nullable
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean lessThan(@SqlType("unknown") @Nullable Void left, @SqlType("unknown") @Nullable Void right)
+    {
+        return null;
+    }
+
+    @ScalarOperator(LESS_THAN_OR_EQUAL)
+    @Nullable
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean lessThanOrEqual(@SqlType("unknown") @Nullable Void left, @SqlType("unknown") @Nullable Void right)
+    {
+        return null;
+    }
+
+    @ScalarOperator(GREATER_THAN)
+    @Nullable
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean greaterThan(@SqlType("unknown") @Nullable Void left, @SqlType("unknown") @Nullable Void right)
+    {
+        return null;
+    }
+
+    @ScalarOperator(GREATER_THAN_OR_EQUAL)
+    @Nullable
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean greaterThanOrEqual(@SqlType("unknown") @Nullable Void left, @SqlType("unknown") @Nullable Void right)
+    {
+        return null;
+    }
+
+    @ScalarOperator(BETWEEN)
+    @Nullable
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean between(@SqlType("unknown") @Nullable Void value, @SqlType("unknown") @Nullable Void min, @SqlType("unknown") @Nullable Void max)
+    {
+        return null;
+    }
+
+    @ScalarOperator(HASH_CODE)
+    @Nullable
+    @SqlType(StandardTypes.BIGINT)
+    public static Long hashCode(@SqlType("unknown") @Nullable Void value)
+    {
+        return null;
+    }
+}

--- a/presto-main/src/test/java/com/facebook/presto/type/TestUnknownOperators.java
+++ b/presto-main/src/test/java/com/facebook/presto/type/TestUnknownOperators.java
@@ -1,0 +1,111 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.type;
+
+import com.facebook.presto.operator.scalar.AbstractTestFunctions;
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.facebook.presto.type.UnknownType.UNKNOWN;
+
+public class TestUnknownOperators
+        extends AbstractTestFunctions
+{
+    @Test
+    public void testLiteral()
+            throws Exception
+    {
+        assertFunction("NULL", UNKNOWN, null);
+    }
+
+    @Test
+    public void testEqual()
+            throws Exception
+    {
+        assertFunction("NULL = NULL", BOOLEAN, null);
+    }
+
+    @Test
+    public void testNotEqual()
+            throws Exception
+    {
+        assertFunction("NULL <> NULL", BOOLEAN, null);
+    }
+
+    @Test
+    public void testLessThan()
+            throws Exception
+    {
+        assertFunction("NULL < NULL", BOOLEAN, null);
+    }
+
+    @Test
+    public void testLessThanOrEqual()
+            throws Exception
+    {
+        assertFunction("NULL <= NULL", BOOLEAN, null);
+    }
+
+    @Test
+    public void testGreaterThan()
+            throws Exception
+    {
+        assertFunction("NULL > NULL", BOOLEAN, null);
+    }
+
+    @Test
+    public void testGreaterThanOrEqual()
+            throws Exception
+    {
+        assertFunction("NULL >= NULL", BOOLEAN, null);
+    }
+
+    @Test
+    public void testBetween()
+            throws Exception
+    {
+        assertFunction("NULL BETWEEN NULL AND NULL", BOOLEAN, null);
+    }
+
+    @Test
+    public void testCastToBigint()
+            throws Exception
+    {
+        assertFunction("cast(NULL as bigint)", BIGINT, null);
+    }
+
+    @Test
+    public void testCastToVarchar()
+            throws Exception
+    {
+        assertFunction("cast(NULL as varchar)", VARCHAR, null);
+    }
+
+    @Test
+    public void testCastToDouble()
+            throws Exception
+    {
+        assertFunction("cast(NULL as double)", DOUBLE, null);
+    }
+
+    @Test
+    public void testCastToBoolean()
+            throws Exception
+    {
+        assertFunction("cast(NULL as boolean)", BOOLEAN, null);
+    }
+}

--- a/presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java
+++ b/presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java
@@ -3455,6 +3455,7 @@ public abstract class AbstractTestQueries
     {
         assertQuery("SELECT orderkey FROM orders UNION SELECT custkey FROM orders");
         assertQuery("SELECT 123 UNION DISTINCT SELECT 123 UNION ALL SELECT 123");
+        assertQuery("SELECT NULL UNION SELECT NULL");
 
         // mixed single-node vs fixed vs source-distributed
         assertQuery("SELECT orderkey FROM orders UNION ALL SELECT 123 UNION ALL (SELECT custkey FROM orders GROUP BY custkey)");

